### PR TITLE
Add support for b_ndebug to rust

### LINKED ISSUE #11438: suggestion for some rust cfg handling in meson
At the moment meson doesn't add any cfg entries itself. Generally I think this is the right approach, but meson does have `b_ndebug` as a core options which can fundamentally affect APIs.

To have it better integrated with `bindgen` + rust targets I'd say meson should add a `--cfg ndebug` (or NDEBUG to match C more closely) on its own, so project do not have to deal with this themselves, because let's face it, checking if `NDEBUG` gets set or not isn't done with a trivial if clause in meson as far as I know.

The main reason for this being important is, that projects might flip some symbols between being a proper function or a simple `static inline` stub, so projects might want to gate certain rust code behind a `#[cfg(not(ndebug))]` so they do not have to deal with the `static inline` `bindgen` situation for those symbols.

### PR DESCRIPTION
**Pull request by dcbaker**
Rust has a `debug_assert!()` macro, which is designed to be toggled on the command line. It is on by default in debug builds, and off by default in release builds, in cargo. This matches what meson's b_ndebug does in `if-release` mode. In addition, this fixes an oversight in the rust module where bindgen wouldn't be passed `-DNDEBUG` when `b_ndebug=false`.

Fixes #11438

### CODE REVIEW

File: docs/markdown/snippets/rustc-ndebug.md
Comment (tristan957) [inline docs/markdown/snippets/rustc-ndebug.md:4]: Perhaps add the man page section that assert belongs to? Seems to be 3. Might help make this a little clearer.

In addition `assert` in Rust is a macro, right? Would it be more correct to add the `!`?

**Review by tristan957 — APPROVED** [review]

File: docs/markdown/snippets/rustc-ndebug.md
Comment (LingMan) [inline docs/markdown/snippets/rustc-ndebug.md:3]: Technically it controls the `debug_assertions` config option. That in turn does control the `debug_assert!()` macro, but developers are also free to put `#[cfg(debug_assertions)]` elsewhere to have debugging code in development but not in production builds.

https://doc.rust-lang.org/reference/conditional-compilation.html#debug_assertions

File: mesonbuild/compilers/compilers.py
Comment (LingMan) [inline mesonbuild/compilers/compilers.py:1049]: s/assertsion/assertions/?

File: docs/markdown/snippets/rustc-ndebug.md
Comment (dcbaker) [reply docs/markdown/snippets/rustc-ndebug.md]: Is the updated snippet better? I actually didn't know that it controled a `cfg()`, that actually makes it really useful since we can then ensure that C and Rust have the same behavior

File: mesonbuild/modules/rust.py
Comment (tristan957) [inline mesonbuild/modules/rust.py:202]: Yes it is!

File: docs/markdown/snippets/rustc-ndebug.md
Comment (LingMan) [reply docs/markdown/snippets/rustc-ndebug.md]: LGTM

File: mesonbuild/compilers/compilers.py
Comment (bruchar1) [inline mesonbuild/compilers/compilers.py:337]: Wouldn't `is_debug_disabled` or something similar be a more appropriate name, since this function do not *disable* anything...

Comment (dcbaker) [reply mesonbuild/compilers/compilers.py:337]: yeah, that makes sense.

Comment (dcbaker) [reply mesonbuild/compilers/compilers.py]: I went with `are_assertions_disabled`, since this is really checking for assertions, and not debug builds.

**Comment by karolherbst** [comment]
any progress on this or any way I can help out with that?

**Comment by tristan957** [comment]
Pretty sure it is just missing a review from a maintainer. Otherwise, this looks sweet to me!

**Comment by karolherbst** [comment]
@jpakkane mind reviewing this? @dcbaker or can this land anyway? Kinda waiting on this to land more stuff on the mesa side.

**Pull request merged**
